### DISCUSSION TITLE
Deployed to AWS - 502 Bad Gateway

### DISCUSSION THREAD

**Discussion by AndryHTC** [serverpod/serverpod#445] [Q&A] [answered]
We have followed every single step of the procedure but at the end of the day... Testing the deployment as suggested https://docs.serverpod.dev/deployments/deploying-to-aws#test-your-deployment we get a **502 Bad Gateway** for all the endpoints.

Same problem with native AWS URL instead of DNS Domains... The connection to the database works, even with the subdomain database.[domain].[*]

**Comment by tusharbhambere** [comment]
Also Check this video https://youtu.be/zx2eh3t_98Q

**Comment by AndryHTC** [comment]
We have made the exact same steps

**Comment by yahu1031** [comment]
Maybe you have missed something buddy. [My API](https://api.minnu.info) works fine.

**Comment by AmoonPod** [comment]
> Maybe you have missed something buddy. [My API](https://api.minnu.info) works fine.


Which OS are you currently using? Amazon Linux 2? Ubuntu?

**Comment by yahu1031** [comment]
@AndryHTC Okie So the bad gateway I got too but when I change the config file and deploy. Can I know What are you trying to access from the url? Like `api.*` or `app.*`?

**Comment by yahu1031** [comment]
@AmoonPod 
> Which OS are you currently using? Amazon Linux 2? Ubuntu?

Ubuntu is the default I guess. I don't specify anything.

**Comment by AndryHTC** [comment, marked as answer]
The problem was that we required in `pubspec.yaml` a different Dart SDK, and both `aws/terraform/init-script.sh` and the `.github/workflows/deployment-aws.yml` install a fixed SDK, and it does not take the same you define in the `pubspec.yaml`.

We actually had the entire infrastructure already installed, so @AmoonPod solved the issue by upgrading the Dart SDK manually in the EC2 instance through the Session Manager console

EDIT:
The permanent solution is to get the minimum version you choose from the `pubspec.yaml` and to put it in the other scripts `aws/terraform/init-script.sh` and `.github/workflows/deployment-aws.yml`. Then re-run the Terraform commands to avoid the regression of the problem in case you will need to terminate and re-start the instance in the future.

**Reply by yahu1031** [reply]
@AndryHTC @AmoonPod
Could you please be a little specific, like where should be changed and what?

Cause
> The problem was that we required in `pubspec.yaml` a different Dart SDK, and both `aws/terraform/init-script.sh` and the `.github/workflows/deployment-aws.yml` install a fixed SDK, and it does not take the same you define in the `pubspec.yaml`.

This actually doesn't matter cause the pubspec will take the range of `x.x.x` - `y.y.y` versions.

> solved the issue by upgrading the Dart SDK manually in the EC2 instance through the Session Manager console

Can you please guide me how to do it.

**Reply by AndryHTC** [reply]
@yahu1031 for upgrading manually the instance you can connect inside the AWS EC2 dashboard. If you chose your instance and click connect will appear terminal of that instance, then just update it as you do on your local machine.

Please share what you have done to the pubspec Dart SDK version for future reference.

The best way I found is to get the minimum version you choose from the `pubspec.yaml` and to put it in the other scripts `aws/terraform/init-script.sh` and `.github/workflows/deployment-aws.yml`. Then re-run the Terraform commands (the manual installation in the EC2 terminal will be replaced in case you will need to terminate and re-start the instance in the future, causing the same problems).

@vlidholt what is the minimum Dart SDK that Serverpod Server needs to run without problems? Does it match the default one when you create the serverpod project 2.18.0?

**Reply by yahu1031** [reply]
@AndryHTC I think min version is `2.17.0` where `2.18.0` has a lot of other upgrades. So I would recommend upgrading the pubspec version itself.

> Please share what you have done to the pubspec Dart SDK version for future reference.

Sadly I've cleaned up my whole AWS 😢. So it wasn't hosted now. Waiting for certificate approval rn.

**Reply by layerzzzio** [reply]
I got the same error and @AndryHTC solution worked for me. 
Let's start by describing my Dart/Flutter/Serverpod settings:
- serverpod 1.1.0
- flutter (web) 3.7.12
- dart 2.19.6

Note: with that configuration, my Flutter web app runs perfectly locally.

# Objective
Changing dart version from 2.18.1 to 2.19.6 in order to fix the Bad Gateway 502 error
> Note: it is possible the 502 error you encounter is not due to the version discrepancies. But in my case, this was the reason (I didn't know until I tried the solution).

# Solution
## 1) Modify dart version in your EC2 instance
Log in to your AWS console and type EC2 in the search bar:
<img width="1073" alt="image" src="https://user-images.githubusercontent.com/98493964/235116274-f59a749b-67c7-468b-83bb-4e28e4ce95e0.png">

Click on the EC2 service and click on Instance and then on Instance ID:
![image](https://user-images.githubusercontent.com/98493964/235116911-cec78dc6-513e-449b-993b-48d87af78b3d.png)

Now you need to connect to your machine so you can modify the version manually:
 - click first on Connect
<img width="1681" alt="image" src="https://user-images.githubusercontent.com/98493964/235117267-e3862bb5-b2be-4c48-b4fb-4677e5c54952.png">

- Choose your method (EC2 instance connect is recommended but do as you want here):
<img width="866" alt="image" src="https://user-images.githubusercontent.com/98493964/235117696-820ee8de-d344-4e35-9f26-e2fc97790fec.png">

- you should see a terminal. In this terminal run the following script (that you can get from `init-script.sh`):
> Note 1: I changed the version to 2.19.6 in the URL
> Note 2: you might need to run the last command with sudo (but normally the export is already done)
```
echo "Installing dart"
wget -q https://storage.googleapis.com/dart-archive/channels/stable/release/2.19.6/sdk/dartsdk-linux-x64-release.zip
unzip -q dartsdk-linux-x64-release.zip
sudo mv dart-sdk/ /usr/lib/dart/
sudo chmod -R 755 /usr/lib/dart/
echo 'export PATH="$PATH:/usr/lib/dart/bin"' >> /etc/profile.d/script.sh
```

- Then I restarted my instance (normally not needed but I still did it - the website was not LIVE at that time)
> WARNING 1: don't do that while users are using your website because it turns the webserver off. They won't have access to the website anymore.
> WARNING 2: doing so scrapped the serverpod tables in the Postgres database. I recreated them by executing the `tables-serverpod.pgsql` script with Postico

- Connect again to your machine and in the terminal type `dart --version`. You should see version `2.19.6` or whatever your version should be.

## 2) Modify dart version in init-script.sh
So in `<YOUR_PROJECT>_server/aws/terraform/init-script.sh`, I changed `wget -q https://storage.googleapis.com/dart-archive/channels/stable/release/2.18.1/sdk/dartsdk-linux-x64-release.zip`by `wget -q https://storage.googleapis.com/dart-archive/channels/stable/release/2.19.6/sdk/dartsdk-linux-x64-release.zip`

## 3) Modify dart version in deployment-aws.yml
Changed:
```
      - name: Setup Dart SDK
        uses: dart-lang/setup-dart@v1.3
        with:
          sdk: 2.18.1
```
to
```
      - name: Setup Dart SDK
        uses: dart-lang/setup-dart@v1.3
        with:
          sdk: 2.19.6
```

## 4) Re-execute terraform commands
> Note: I am not familiar with Terraform so the first command might not be necessary. But it doesn't hurt to do it.

Go to the following directory `<YOUR_PROJECT>_server/aws/terraform`and execute:
```
terraform init
terraform apply
```

## 5) Commit all this to GitHub and manually trigger an AWS Deployment.

Result: the Bad Gateway 502 error disappeared!
> Final note: hopefully all the steps chain up well and you get your error fixed. If not, you might have to tweak a few things on your own. But most of it is in the above explanations. Credits to @AndryHTC 🙏

**Reply by husain-nahar** [reply]
Thanx for the info.
I am stuck at whenever I run sudo mv dart-sdk/ /usr/lib/dart/ it says cannot move ‘dart-sdk/’ to ‘/usr/lib/dart/dart-sdk’: File exists. 
Also if I try to rm dart-sdk inside /usr/lib/dart it says cannot remove. And then I skip this step as it is not working and 
go for sudo chmod -R 755 /usr/lib/dart/ at last of all the version of dart shows the same on my linux machine on ec2. 
Please help detailing it more on how to update dart SDK on linux.  As I am also dealing with BAD GATEWAY 502 error. 
NOTE: Current version on ec2 linux is showing 3.1.0, if I do opposite and go to use this version inside my pubspec.yaml file of server it says you have to have version above 3..2.0 as it is bound with constraints of Serverpod latest version.

**Reply by husain-nahar** [reply]
```
echo "Installing dart"
wget -q https://storage.googleapis.com/dart-archive/channels/stable/release/2.19.6/sdk/dartsdk-linux-x64-release.zip
unzip -q dartsdk-linux-x64-release.zip
sudo mv dart-sdk/ /usr/lib/dart/
sudo chmod -R 755 /usr/lib/dart/
echo 'export PATH="$PATH:/usr/lib/dart/bin"' >> /etc/profile.d/script.sh
```
Doing above I am not succeeding in updating my dart sdk version from 3.1.0 to 3.5.0(latest).
After all doing, it still shows 3.1.0 only on my was ec2 linux.

Also my **echo $PATH** says this:
`/usr/local/bin:/usr/bin:/usr/local/sbin:/usr/sbin:/usr/lib/dart/bin:/home/ec2-user/.local/bin:/home/ec2-user/bin:/usr/lib/dart/bin:/home/ec2-user/.local/bin:/home/ec2-user/bin:/usr/lib/dart/bin:/home/ec2-user/.local/bin:/home/ec2-user/bin:/usr/lib/dart/bin:/home/ec2-user/.local/bin:/home/ec2-user/bin`

Now my server has stopped responding at 502 bad gateway.
The only way for now to make it run is downgrading Serverpod version to 2.0.0 which supports dart sdk version 3.1.0 same as on my ec2 linux machine.

When I try updating dart using this [](https://dart.dev/get-dart#install).
In the first step itself it shows error saying apt-get command not found.

@vlidholt  please address this issue.
Now I am stuck in a scenario where my flutter code is written supporting the latest serverpod 2.0.2 and dart sdk 3.5.0 but my server linux is still using 2.0.0 with support of dart sdk 3.1.0.

**Reply by husain-nahar** [reply]
Ok I fixed the issue.
the code you suggested as last step is:
`echo 'export PATH="$PATH:/usr/lib/dart/bin"' >> /etc/profile.d/script.sh`, 
rather it should be:
`echo 'export PATH="/usr/lib/dart/dart-sdk/bin:$PATH"' | sudo tee -a /etc/profile.d/script.sh`.
After this it should say:
`export PATH="/usr/lib/dart/dart-sdk/bin:$PATH"`.
If this comes that means you have successfully updated dart-sdk, also logout and reconnect ec2 server again n type `dart --version`, you should see your desired version.
NOTE: if `sudo mv dart-sdk/ /usr/lib/dart/` says: dart-sdk file exists already. please do the following:
1. `sudo rm -rf /usr/lib/dart/dart-sdk` n then go for `sudo mv dart-sdk/ /usr/lib/dart/`.

**Comment by vlidholt** [comment]
Converted this to a question so that more people potentially can get help from it.

**Reply by dongri727** [reply]
I’m still stuck in 502 bad gateway after these steps above
Any additional good advice ?

Serverpod 1.1.1
Flutter 3.13.9
Dart 3.1.5

I could terraform apply
I could  deploy to AWS

 502 on Chrome, Safari and Brave

**Reply by AndryHTC** [reply]
Any runtime error that prevents Serverpod from running will cause that error. Try to strip down and decompose your app piece by piece

**Reply by jblankenship5** [reply]
@AndryHTC  when deploying to GCP I was able to view the logs from serverpod attempting to run, how can i do the same with aws?

**Comment by AndryHTC** [comment]
> @AndryHTC when deploying to GCP I was able to view the logs from serverpod attempting to run, how can i do the same with aws?

I usually do a `cat serverpod.log` at the default **ec2-user** directory

**Reply by husain-nahar** [reply]
@AndryHTC man I updated my ec2 linux dart sdk to 3.5.0 and also my pubspec has the same. when I was uploading with 3.1.0 of dart it was all running fine. but after tune-ing in to 3.5.0. that 502 bad gateway appeared again. Any help on this is highly appreciated.

**Reply by AndryHTC** [reply]
You need to be sure you update the version in all places. Pay attention also to the Terraform initialization script, it's used every time the ec2 instance is rebooted

**Reply by husain-nahar** [reply]
you mean init-script.sh inside terraform folder?
This is what it is there:

# Install Dart

```
echo "Installing dart"

wget -q https://storage.googleapis.com/dart-archive/channels/stable/release/3.5.0/sdk/dartsdk-linux-x64-release.zip

unzip -q dartsdk-linux-x64-release.zip

sudo mv dart-sdk/ /usr/lib/dart/

sudo chmod -R 755 /usr/lib/dart/

echo 'export PATH="$PATH:/usr/lib/dart/bin"' >> /etc/profile.d/script.sh

```

**Comment by hidayet-poscu** [comment]
aws 
[ec2-user@ip-10-0-6-131 ~]$ dart --version
**Dart SDK version: 3.5.0** (stable) (Tue Jul 30 02:17:59 2024 -0700) on "linux_x64"
[ec2-user@ip-10-0-6-131 ~]$    

vscode
hidayet@192 terraform % dart --version
**Dart SDK version: 3.5.0** (stable) (Tue Jul 30 02:17:59 2024 -0700) on "macos_arm64"
  Flutter 3.24.0 
 serverpod: 2.0.2

conclusion:  502 Bad Gateway

please help
